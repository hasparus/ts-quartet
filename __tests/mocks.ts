import {
  FunctionSchema,
  IVariantSchema,
  Schema,
  IObjectSchema
} from "../types";
import { v } from "../index";

export const primitives: any[] = [
  undefined,
  null,
  true,
  false,
  0,
  1,
  NaN,
  Infinity,
  -Infinity,
  "",
  "Andrew",
  Symbol.for("test")
];

export const funcSchema: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`
});
export const funcSchemaWithNot: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`
});
export const funcSchemaWithHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithNotHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divide !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divider !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});

export const funcSchemas: FunctionSchema[] = [
  funcSchema,
  funcSchemaWithNot,
  funcSchemaWithHandleError,
  funcSchemaWithNotHandleError,
  funcSchemaWithPrepare,
  funcSchemaWithNotWithPrepare,
  funcSchemaWithHandleErrorWithPrepare,
  funcSchemaWithNotHandleErrorWithPrepare
];

const has = (obj: any, key: string | number) => {
  return Object.prototype.hasOwnProperty.call(obj, key);
};

export function* objects(
  level: number = 0,
  maxLevel: number = 2
): Iterable<object | any[]> {
  if (level > maxLevel) return;
  yield {};
  yield [];
  for (const obj of objects(level + 1, maxLevel)) {
    if (has(obj, "push")) {
      for (const primitive of primitives) {
        yield [...(obj as any), primitive];
      }
      for (const o of objects(level + 1, maxLevel)) {
        yield [...(obj as any), o];
      }
    } else {
      for (const primitive of primitives) {
        if (!has(obj, "even")) {
          yield { ...obj, even: primitive };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: primitive };
        }
      }
      for (const o of objects(level + 1, maxLevel)) {
        if (!has(obj, "even")) {
          yield { ...obj, even: o };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: o };
        }
        if (!has(obj, "nested")) {
          yield { ...obj, nested: o };
        }
      }
    }
  }
}

export function* variants(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IVariantSchema> {
  if (level > maxLevel) {
    return;
  }
  yield [];
  for (const variant of variants(level + 1)) {
    for (const primitive of primitives) {
      yield [...variant, primitive];
    }
    for (const funcSchema of funcSchemas) {
      yield [...variant, funcSchema];
    }
    for (const obj of objectSchemas(level + 1, maxLevel)) {
      yield [...variant, obj];
    }
    for (const variant2 of variants(level + 1, maxLevel)) {
      yield [...variant, variant2];
    }
  }
}

export function* objectSchemas(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IObjectSchema> {
  if (level > maxLevel) return;
  yield {};
  for (const obj of objectSchemas(level + 1, maxLevel)) {
    if (!has(obj, "even")) {
      for (const primitive of primitives) {
        yield { ...obj, even: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, even: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, even: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), even: obj };
      }
    }
    if (!has(obj, "nested")) {
      for (const primitive of primitives) {
        yield { ...obj, nested: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, nested: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, nested: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), nested: obj };
      }
    }
    if (!has(obj, v.rest)) {
      for (const primitive of primitives) {
        yield { ...obj, [v.rest]: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, [v.rest]: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, [v.rest]: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), [v.rest]: obj };
      }
    }
  }
}

export function* schemas(maxLevel: number): Iterable<Schema> {
  yield* primitives[Symbol.iterator]();
  yield* funcSchemas[Symbol.iterator]();
  yield* objectSchemas(0, maxLevel);
  yield* variants(0, maxLevel);
}

export function* values(maxLevel: number) {
  yield* primitives[Symbol.iterator]();
  yield* objects(0, maxLevel);
}
