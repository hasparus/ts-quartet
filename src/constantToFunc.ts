import { getKeyAccessor } from "./getKeyAccessor";
import { toContext } from "./toContext";
import { ConstantSchema, FunctionSchema } from "./types";

export function constantToFunc(c: ConstantSchema): FunctionSchema {
  if (c === undefined) {
    return () => ({
      check: id => `${id} === undefined`,
      not: id => `${id} !== undefined`
    });
  }
  if (c === null) {
    return () => ({
      check: id => `${id} === null`,
      not: id => `${id} !== null`
    });
  }
  if (typeof c === "boolean") {
    return () => ({
      check: id => `${id} === ${c}`,
      not: id => `${id} !== ${c}`
    });
  }
  if (typeof c === "string") {
    return () => ({
      check: id => `${id} === ${JSON.stringify(c)}`,
      not: id => `${id} !== ${JSON.stringify(c)}`
    });
  }
  if (typeof c === "number") {
    return Number.isNaN(c)
      ? () => ({
          check: id => `Number.isNaN(${id})`,
          not: id => `!Number.isNaN(${id})`
        })
      : () => ({
          check: id => `${id} === ${c}`,
          not: id => `${id} !== ${c}`
        });
  }
  const [cId, prepare] = toContext("constant", c);
  const constantAccessor = getKeyAccessor(cId);
  return () => ({
    check: (id, ctx) => `${id} === ${ctx}${constantAccessor}`,
    not: (id, ctx) => `${id} !== ${ctx}${constantAccessor}`,
    prepare
  });
}
