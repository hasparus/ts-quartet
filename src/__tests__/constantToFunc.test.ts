import { v } from '../index'
import { constantToFunc } from "../constantToFunc";
import { primitives } from "./mocks";

describe("constantToFunc", () => {
  test("works", () => {
    for (const primitive of primitives) {
      const funcRes = constantToFunc(v, primitive)();
      expect(funcRes.check("value", "validator")).toMatchSnapshot();
      expect(
        funcRes.handleError && funcRes.handleError("value", "validator")
      ).toMatchSnapshot();
      expect(
        funcRes.not && funcRes.not("value", "validator")
      ).toMatchSnapshot();
      expect(
        funcRes.prepare &&
          (() => {
            const ctx: any = {};
            funcRes.prepare(ctx);
            return ctx;
          })()
      ).toMatchSnapshot();
    }
  });
});
