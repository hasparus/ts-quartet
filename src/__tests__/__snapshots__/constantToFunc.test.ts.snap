// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constantToFunc works 1`] = `"value === undefined"`;

exports[`constantToFunc works 2`] = `undefined`;

exports[`constantToFunc works 3`] = `"value !== undefined"`;

exports[`constantToFunc works 4`] = `undefined`;

exports[`constantToFunc works 5`] = `"value === null"`;

exports[`constantToFunc works 6`] = `undefined`;

exports[`constantToFunc works 7`] = `"value !== null"`;

exports[`constantToFunc works 8`] = `undefined`;

exports[`constantToFunc works 9`] = `"value === true"`;

exports[`constantToFunc works 10`] = `undefined`;

exports[`constantToFunc works 11`] = `"value !== true"`;

exports[`constantToFunc works 12`] = `undefined`;

exports[`constantToFunc works 13`] = `"value === false"`;

exports[`constantToFunc works 14`] = `undefined`;

exports[`constantToFunc works 15`] = `"value !== false"`;

exports[`constantToFunc works 16`] = `undefined`;

exports[`constantToFunc works 17`] = `"value === 0"`;

exports[`constantToFunc works 18`] = `undefined`;

exports[`constantToFunc works 19`] = `"value !== 0"`;

exports[`constantToFunc works 20`] = `undefined`;

exports[`constantToFunc works 21`] = `"value === 1"`;

exports[`constantToFunc works 22`] = `undefined`;

exports[`constantToFunc works 23`] = `"value !== 1"`;

exports[`constantToFunc works 24`] = `undefined`;

exports[`constantToFunc works 25`] = `"Number.isNaN(value)"`;

exports[`constantToFunc works 26`] = `undefined`;

exports[`constantToFunc works 27`] = `"!Number.isNaN(value)"`;

exports[`constantToFunc works 28`] = `undefined`;

exports[`constantToFunc works 29`] = `"value === Infinity"`;

exports[`constantToFunc works 30`] = `undefined`;

exports[`constantToFunc works 31`] = `"value !== Infinity"`;

exports[`constantToFunc works 32`] = `undefined`;

exports[`constantToFunc works 33`] = `"value === -Infinity"`;

exports[`constantToFunc works 34`] = `undefined`;

exports[`constantToFunc works 35`] = `"value !== -Infinity"`;

exports[`constantToFunc works 36`] = `undefined`;

exports[`constantToFunc works 37`] = `"value === \\"\\""`;

exports[`constantToFunc works 38`] = `undefined`;

exports[`constantToFunc works 39`] = `"value !== \\"\\""`;

exports[`constantToFunc works 40`] = `undefined`;

exports[`constantToFunc works 41`] = `"value === \\"Andrew\\""`;

exports[`constantToFunc works 42`] = `undefined`;

exports[`constantToFunc works 43`] = `"value !== \\"Andrew\\""`;

exports[`constantToFunc works 44`] = `undefined`;

exports[`constantToFunc works 45`] = `"value === validator.constant"`;

exports[`constantToFunc works 46`] = `undefined`;

exports[`constantToFunc works 47`] = `"value !== validator.constant"`;

exports[`constantToFunc works 48`] = `
Object {
  "constant": Symbol(test),
}
`;
